src/Rule/Path.php

```diff
@@ -82,7 +82,17 @@ public function __invoke(ServerRequestInterface $request, Route $route)
             return false;
         }
 
-        $route->attributes($this->getAttributes($matches, $route->wildcard));
+        $attributes = $this->getAttributes($matches, $route->wildcard);
+
+        foreach ($this->route->tokens as $name => $pattern) {
+            if (is_callable($pattern)) {
+                if (!$pattern($attributes[$name], $route, $request)) {
```

Comment:
I am not sure my thoughts are a valid one. But there can be a problem. The `$request` object is psr-7 which is immutable. So if they change inside the custom function it will not be changed when got from other middleware. Another thing is this looks it can only pass 3 params. I am not sure if there is a need for another though. Does anyone feel it is not enough ?